specs/plasma.md

```diff
@@ -0,0 +1,114 @@
+# Plasma mode
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [DA Storage](#da-storage)
+- [Input Commitment](#input-commitment)
+- [Data Availability Challenge Contract](#data-availability-challenge-contract)
+- [Derivation](#derivation)
+- [Safety and finality](#safety-and-finality)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Using Plasma mode means switching to an offchain data availability provider.
+The protocol guarrantees availability with a challenge contract on L1 where users
+can challenge the availability of input data used to derive the chain.
+Challenging an input commitment means forcing providers to submit the input data on chain
+to prove the data is available during a time period. We call the period during which any party
+can challenge a commitment `challenge_window` and the period during which any party
+can submit the input data to resolve the challenge `resolve_window`.
+If a challenge is not resolved by the time `resolve_window` of L1 blocks have passed,
+chain derivation must be reset, omitting the input data when rederiving therefore causing a reorg.
+
+## DA Storage
+
+Input data is uploaded to the storage layer via plain HTTP calls to a stateless DA storage service.
+This service is horizontally scalable and concerned with replicating the data across prefered storage layers
+such as IPFS or any S3 compatible storage. Input data is content addressed by its hash in the request url
+so responses are easily cachable.
+
+Any DA provider can implement the following endpoints to receive and serve input data:
+
+- `POST /put/<hex_encoded_commitment> -H "Content-Type: application/octet-stream"`
+- `GET /get/<hex_encoded_commitment> -H "Content-Type: application/octet-stream"`
+
+## Input Commitment
+
+The batching and compression of input data remain unchanged. When a batch is ready
+to be submitted to the inbox address, the data is uploaded to the DA storage layer instead, and a
+commitment (keccak256 hash) is submitted as the bacher inbox transaction call data.
+The batcher will not submit a commitment onchain unless input data was successfully stored on the service.
+
+## Data Availability Challenge Contract
+
+Data availability is guaranteed via a permissionless challenge contract on the L1 chain.
+Users have a set number of L1 blocks (AKA `challenge_window`) during which they are able to call
+the `challenge` method of the contract with the following inputs:
+
+- A Commitment type (i.e. keccak256)
```

Comment:
Can you describe how the versioning scheme works in more depth? ie is the version encoded in the first byte of the commitment? And what are the defined versions? We should consider a "reserved range" for forks to use

Reply:
In the current design we weren't planning to include the commitment type in the commitment data posted to the batch inbox itself, but only use it as a parameter in the challenge contract (and make it a system config variable per rollup) - that means the commitment type couldn't change per commitment but only via the system config and we don't have pay for the commitment type bytes in every commitment.

We're currently planning to have a `bytes32` commitment type, which is provided to the `challenge` function as a separate parameter and used as one of the keys for the mapping that stores the challenge data. The `resolve` function would also have this additional parameter and doesn't include the implementation to check whether the provided data corresponds to the challenge, but rather calls out to an implementation contract that is registered in the challenge contract per commitment type. The nice thing with `bytes32` is that we could use things like `keccak256("keccak256")` as commitment type. But it'd also be easy to reduce the size of the commitment type, the main benefit being less calldata cost.

Ie something like this:

```solidity
function challenge(uint256 challengedBlockNumber, bytes32 commitmentType, bytes calldata commitment) public {
  // Check requirements
  // Store the challenge in the `challenges` mapping
}

function resolve(uint256 challengedBlockNumber, bytes32 commitmentType, bytes calldata commitment, bytes calldata inputData) public {
  // Look up challenge in the `challenges`
  // Look up resolver implementation contract for the provided `commitmentType`
  // Call resolver implementation contract for the provided `commitmentType` to verify the inputData
  // Resolve the challenge
}

```

What are your thoughts on this @tynes? If you generally agree with the approach we can add it to the spec here, curious about into how much implementation detail we should already go in this spec.

Reply:
> In the current design we weren't planning to include the commitment type in the commitment data posted to the batch inbox itself, but only use it as a parameter in the challenge contract (and make it a system config variable per rollup) - that means the commitment type couldn't change per commitment but only via the system config and we don't have pay for the commitment type bytes in every commitment.
> 
> We're currently planning to have a `bytes32` commitment type, which is provided to the `challenge` function as a separate parameter and used as one of the keys for the mapping that stores the challenge data. The `resolve` function would also have this additional parameter and doesn't include the implementation to check whether the provided data corresponds to the challenge, but rather calls out to an implementation contract that is registered in the challenge contract per commitment type. The nice thing with `bytes32` is that we could use things like `keccak256("keccak256")` as commitment type. But it'd also be easy to reduce the size of the commitment type, the main benefit being less calldata cost.
> 
> Ie something like this:
> 
> ```solidity
> function challenge(uint256 challengedBlockNumber, bytes32 commitmentType, bytes calldata commitment) public {
>   // Check requirements
>   // Store the challenge in the `challenges` mapping
> }
> 
> function resolve(uint256 challengedBlockNumber, bytes32 commitmentType, bytes calldata commitment, bytes calldata inputData) public {
>   // Look up challenge in the `challenges`
>   // Look up resolver implementation contract for the provided `commitmentType`
>   // Call resolver implementation contract for the provided `commitmentType` to verify the inputData
>   // Resolve the challenge
> }
> ```
> 
> What are your thoughts on this @tynes? If you generally agree with the approach we can add it to the spec here, curious about into how much implementation detail we should already go in this spec.

My current understanding is that the goal here is to ship something that works in the specific case that you are building for and then later make it more flexible. Apologies for the communication back and forths but perhaps we cut the concept of version for now and then in a hard fork we work with other DA teams to build out an versioned interface that works for everybody.

For this upgrade, my intuition is that we want the version byte in the first byte of the data posted to the chain and the version indicates a schema for the serialized commitment (size, multiple fields). There are likely things that I don't understand right now that may make this design space bad.

Reply:
If we have spoken to other teams and gotten the OK that the interface as defined will work for them, then I am fine moving forward with it